### ISSUE TITLE
Got My 3 Numbers Banned!

### ISSUE THREAD

**Issue by AmitGamingKing** [danielcardeenas/whatsapp-framework#156]
Hello, As I'm totally new here I just started using it! After some days my interest level get maximum! At first I just used inbuilt modules while I was learning Python... And my 1st number get banned in 27-28 hours! Used second number get banned instantly after registration! Used 3rd number  (My friend's) created many own modules got success in every module! Number was running fine until 4-5 days and guess what next? *BANNED AGAIN* Now please suggest me how to not get banned! I've already read the "How To Not Get Banned" section in repo. I registered using android env! Thanks in advance!

**Comment by ngethe** [comment]
Do not send messages to contacts that are not in your contact list inside config.py (haven't been sync)
Send messages only to whom have messaged you before (except groups)
Syncs all the contacts in the groups you belong to by adding them into your contact list inside config.py
Do not register several numbers / login to whatsapp with the same device (same IP address) or at the same time. this will lead to the IP address being blacklisted.

**Comment by elchory96** [comment]
Look, what the above says is true, you must follow these steps, but if the ban is to register the number with yowsup-cli, I tell you that this problem also happened to me, what I did was first record it on a cell phone common and current, adjusting once the profile image and status, after more or less one day of registered on the cell phone, there is if you register with yowsup-cli.

**Comment by AmitGamingKing** [comment]
I think that's the mistake I did! I didn't add the contact list in config.py so possibly I banned for that reason! And I registered through yowsup-cli is there any other way to register? Like if I register through my phone and then insert the 6 digit code directly in yowsup-cli is that possible? And at last also give me some tips to prevent ban as I'm buying a new number...

**Comment by elchory96** [comment]
That's why I tell you, the first thing I do when buying a new sim, is to register the number in whatsapp but from a current phone, after one day of being registered in the phone, I migrate it to the linux server, and registering it again with yowsup-cli.

So far it has worked for me, and also added validations that say in the [wiki](https://github.com/danielcardeenas/whatsapp-framework/wiki/Do-not-get-banned)

**Comment by AmitGamingKing** [comment]
Ok so first I'll buy a SIM Card then I'll register WhatsApp on my normal phone first for one day and then I'll use yowsup-cli registration process and before that I'll add all contacts in config.py! Is that ok? But in Wiki at the last line it says do not register with yowsup-cli what does this sentence mean??

**Comment by JosemariC** [comment]
Hi, I think that refers to not using yowsup-cli to send messages. These should be sent only from the whatsapp-framwork (the bot). It is also very important (to avoid blocking) not to send messages to several recipients at the same time. You must queue messages (with random delays) as a human would.

**Comment by AmitGamingKing** [comment]
Yeah I never used used yowsup-cli to send messages I only used the bot! And this time I'll write a delay line in all scripts to make it stable and safe! By the way why WhatsApp don't ban DUTA Services if anyone heard about DUTA?

**Comment by AmitGamingKing** [comment]
Banned my 4th number too yesterday! Server was running and suddenly I got auth error! This number worked fine for 25 days! Even I followed all rules and the number I only shared with me and my friend! Still got banned I've filled contact list in config.py who are using this bot! Changing number after each ban is not possible so it makes this library useless! One doubt I'm having I used ```mac.contact_picture_from``` feature in my bot and used random numbers for DPs is it safe?

**Comment by MarAvFe** [comment]
All of you who are filling the contact list, are you doing so dynamically or manually?

I thought of, receiving a message from a new recipient, add it to the file, and rebooting the bot. I haven't understood well where and how the framework synchs contacts. 

Also, that list could be read from a database and synching constantly.

**Comment by elchory96** [comment]
Hello, I asked a question a while ago about this topic and we could solve it, please check it out.

[issue#129](https://github.com/danielcardeenas/whatsapp-framework/issues/129)

**Comment by AmitGamingKing** [comment]
@MarAvFe what do you mean by dynamically? I've already added the numbers in ``config.py`` who are going to use this bot and they were only 4 numbers including me & my 3 friends! And I kept that number secret still I my numbers are getting ban!

**Comment by MarAvFe** [comment]
Well.. Lets say I want a calendar bot to my school. Instead of adding 500 numbers with my keyboard, I'd like to post the number on some board and everybody sent "Hi bot!" and the bot detects that number, reads it's number and then adds it to a db or to the heap at least. 

But elchory96 is right, he created a json for the bot. From there on, it'll only be read that json and save it in some mongo or so.

**Comment by AmitGamingKing** [comment]
That's the another thing... My question was different I don't want to know how can I add numbers automatically in config instead I can do it manually! My question was why my bot is getting banned even I followed all precautions!

**Comment by MarAvFe** [comment]
#175 Check my issue. That might be a reason.

**Comment by elchory96** [comment]
A solution not to be banned might be what MarAvFe says, I have not tried it, but it can work.

I have been using my bot for almost 4 months, and they have not yet banned the number.

Something that you have to keep in mind is that you have to make a list in queue of the messages that you are receiving, because currently the framework responds at the same time to all the users that send you a message, I have to place a queue, to know when finish sending a message to a user, can go with the next, because sending a message for example to 5 people at the same time, it would not be very human.

**Comment by AmitGamingKing** [comment]
Yeah but doing this thing is pretty much hard for me! Ok I want to ask you some questions please respond to it:

- What type of modules are you using? Simple like just hi/hello/wassup or advance services like knowing weather and all other stuffs by importing another packages?

- Are you using any prefix for your messages?

- Is your bot private or publically visible?

- Did you use direct new registration or first registered on WhatsApp then Yowsup?

- What type of device you're hosting your bot like local PC/laptop or any VPS?

Kindly answer these questions it will help a lot!!!

**Comment by elchory96** [comment]
about your questions ... 

(1). My bot is dedicated to obtain services of any kind or to look for a house in a location by map or by means of city or neighborhood. 

(2). No, I do not use any prefix in messages, I have a database with a table where I keep the questions that should be asked to the user and the possible answers to each question. 

(3). The bot is public, everyone who has the number will answer and at the same time save the number in contacts automatically, currently they are talking on average 40 users a day. 

(4). The registration I did first in a physical device, to be able to change profile picture and contact information, after that spent a day or so, I registered it with the Yowsup library. 

(5). While testing it, I created it and registered it on a Mac computer, after it was ready I passed it to a Linux server on Amazon web services.

**Comment by AmitGamingKing** [comment]
Great! Thanks for answering my questions! I'm also hosting my bot on AWS EC2 but I'm registering directly on EC2 instance... I bought a SIM and register on physical device and it's been almost 1 week! Now today I'm gonna register on Yowsup let's hope for the best guys! Thank you so much for your quick responses and help! 😊

**Comment by JoseMPinyol** [comment]
> Awesome that's what I was looking for! Possibly this might be my banning reason! As everyone was rapidly giving messages to bot and another reason might be command messages as normal humans don't talk in command language with a prefix `!` so WhatsApp automated system detects that this may be not a human and banned it automatically!. 
Do not worry for that (They can't use the message content in order to ban you).
Maybe you should look at the following link.
https://techcrunch.com/2017/02/02/how-whatsapp-is-fighting-spam-after-its-encryption-rollout/
I'm still struggling with the same problem. Good look.

**Comment by ngethe** [comment]
also noted IP that are not from my country are banned (e.g. when I use aws server that is not in my country I get banned while other servers register as needed)

**Comment by g33klord** [comment]
today two of my bots stopped working. One was running on the local system and another one on digitalocean instance. Bot fail to start again with the message:

```
[Whatsapp] Mac started

Auth Error, reason not-authorized
```
Surprisingly, I was able to run  `yowsup-cli registration --requestcode sms --phone 91XXXXXXXXX --cc 91 -E android` command and get a new password. 

I spun up a new digitalocean instance and registered with a new number. Registration was successful but again bot fails to start with  `Auth Error, reason not-authorized` message.

Is my number is blocked? Why I was able to register with the new number and not start the bot?
